### ISSUE TITLE
[Feature Request] A different way to optimize the algorithm

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs4anki#339]
**Which module is related to your feature request?**
Optimizer

**Is your feature request related to a problem? Please describe.**
While our attempts to use the R-matrix to improve accuracy by combining theoretical R with R derived from the matrix have failed, I wonder if it's possible to use it for optimization. We could minimize the difference between theoretical R and its corresponding R-matrix entry.

**Describe the solution you'd like**
The following loss function should be used:
`-torch.log(1-torch.abs(predictions-labels))`, where predictions are theoretical values of R, and labels are R-matrix entries
I sugegst the above loss because BCE loss is very unintuitive. It can be above 0 even when predictions **exactly** match labels.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/b3fe8cef-26db-434e-82a0-622a0d0bb2a8)
When both x and y are exactly the same, any reasonable loss function should be 0. Also, making the labels **not** the same as predictions **improves** the BCE loss. This is the exact opposite of what one would expect. If the difference between labels and predictions increased, the loss function shouldn't increase.

**Additional context**
Frankly, I don't have enough knowledge of pytorch to help you, you will have to figure it out on your own. Theoreticallly, this seems doable. But it's possible that it's somehow incompatible with gradient descent or backpropagation.

**Comment by Expertium** [comment]
@user1823 at some point mentioned Goodhart's law: "When a proxy for some value becomes the target of optimization pressure, the proxy will cease to be a good proxy". Broadly speaking, I can think of 2 very different versions of it: Adversarial Goodhart and Correlational Goodhart. Adversarial Goodhart isn't something we have to be worried about since it involves people deliberately trying to cheat the proxy (for example, children and school grades). But we do have to worry about Correlational Goodhart. Correlational Goodhart occurs when the proxy P and the real metric M that we actually care about aren't perfectly in sync; in other words, their correlation coefficient is <1. In that case, optimizing P won't always improve M and might even make it worse in extreme cases. And in our case, the correlation between BCE loss (the proxy) and RMSE (what we actually care about) is surprisingly low. Around 0.4-0.5, although it varies a lot (I just measure correlation between logloss and RMSE for some version of the algorithm, and since I did a ton of tests for lots of versions, I have a lot of data points). I often see that logloss barely changes or even slightly increases, but RMSE goes down.
The reason I'm talking about all of this is that I think we might just have a bad proxy, so we need a different proxy that is more strongly correlated with what we actually want, which is low RMSE. If optimization based on minimizing the difference between theoretical R and measured R is possible, it could be a much better proxy and `maybe` result in dramatically better RMSE.

**Comment by L-M-Sherlock** [comment]
I implement a simple version:

https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/candidate/grade-derived_R.ipynb

Here is the changes:

Calculate the Grade-derived R:

<img width="1057" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/5ccff425-1edc-4ec7-a320-2dc37f394a80">

Use Grade-derived R as the label:

<img width="772" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/b7619440-fcf7-4d78-ad8d-052b129a3c66">

Use WozLoss as the loss function:

<img width="976" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/066c6400-5ae2-4ecc-81fe-67762627cac3">

**Comment by Expertium** [comment]
No, that's not what I want. Conceptually, this isn't very different since we're still asking the algorithm to predict the outcome of a _single_ review. I want the algorithm to predict _group averages_.
I recommend modifying it by adding 4 weights and making weighted R, like I described in the other issue. And bring the old loss back for this file. Basically, I'm asking you to use this file for grade-derived R, not to test the new optimization procedure.
Again, just to make sure that you understand the key idea. Instead of making the algorithm predict the outcome of individual reviews, we group reviews (using the R-matrix) and make the algorithm predict average R within each group.

**Comment by L-M-Sherlock** [comment]
I don't get your point. Could you provide a several steps that you want to implement in the optimization procedure?

**Comment by Expertium** [comment]
1. Algorithm makes predictions for individual reviews.
2. R-matrix is created by grouping reviews according to their S, D and predicted R.
3. Within each category average measured R is calculated. You have implemented that already in a different file.
4. We take 2 values, theoretical R and the corresponding measured R, and pass them to the loss function. In other words, the loss function does not deal with outcomes of individual reviews.
5. It will also have to somehow be weighted based on the count for each entry.
See the image below. This is what we pass to the loss function.
![20230707_125159](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/bea3b5ae-7976-45c6-8e90-15691099f161)

Theoretical R, measured R, and counts as weights. I don't know if that's even possible while using batches though.

**Comment by L-M-Sherlock** [comment]
I get it. I will try to implement it.

What about Grade-derived R?

**Comment by Expertium** [comment]
Note that it means that the scheduler will receive 8 new parameters, but only 4 of them - the weights - require gradient descent to be optimized.

**Comment by L-M-Sherlock** [comment]
> * Algorithm makes predictions for individual reviews.
> * R-matrix is created by grouping reviews according to their S, D and predicted R.
> * Within each category average measured R is calculated. You have implemented that already in a different file.
> * We take 2 values, theoretical R and the corresponding measured R, and pass them to the loss function. In other words, the loss function does not deal with outcomes of individual reviews.

Here are some technical difficulties:

1. These optimization procedure is incompatible with mini-batch. The batch sampler doesn't ensure sampling the individual reviews from the same group in R-matrix. So it requires refactor the sampler.
2. The grouping varies over epoch because R-matrix is calculated from a set of parameters. The parameters are changing when training. So the sampler should be built dynamically.

And I doubt the advantage of this way to optimize the algorithm. With mini-batch, the algorithm are also minimizing the prediction error within each batch.

**Comment by Expertium** [comment]
I would like you to try nonetheless.
